File: problemas/haskell/NHAY.hs
import           Data.List (intercalate)

main :: IO ()
main = interact $ unlines . map proccess . parse . lines
  where
    parse:: [String] -> [(String, String, String)]
    parse (s1:s2:s3:ss) = (s1, s2, s3) : parse ss
    parse _             = []

    proccess :: (String, String, String) -> String
    proccess (patternLenght, pattern, text) =
      let occurrences = matches pattern text
          m = read patternLenght :: Int
      in if null occurrences
            then "\n"
            else intercalate "\n" $ show <$> (map (subtract m) occurrences)

data Rep a
  = Null
  | Node a (Rep a) (Rep a)
  deriving (Show)

matches :: (Eq a) => [a] -> [a] -> [Int]
matches ws = map fst . filter (ok . snd) . scanl step (0, root)
  where
    ok (Node vs l r) = null vs
    step (n, t) x = (n + 1, op t x)
    op Null x = root
    op (Node [] l r) x = op l x
    op (Node (v:vs) l r) x =
      if v == x
        then r
        else op l x
    root = grep Null ws
    grep l []     = Node [] l Null
    grep l (v:vs) = Node (v : vs) (next l v) (grep (op l v) vs)
    next Null x = Null
    next (Node [] l r) x = Node [] l r
    next (Node (v:vs) l r) x =
      if v == x
        then next l x
        else Node (v : vs) l r

File: problemas/cpp/NHAY.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> compute_prefix_function(const string &pattern) {
  int m = pattern.length();
  vector<int> pi(m);
  pi[0] = 0;
  int k = 0;
  for (int q = 1; q < m; q++) {
    while (k > 0 && pattern[k] != pattern[q])
      k = pi[k - 1];
    if (pattern[k] == pattern[q])
      k++;
    pi[q] = k;
  }
  return pi;
}

vector<int> kmp(const string &text, const string &pattern) {
  vector<int> matches;
  int n = text.size();
  int m = pattern.size();
  vector<int> pi = compute_prefix_function(pattern);
  int q = 0;
  for (int i = 0; i < n; i++) {
    while (q > 0 && pattern[q] != text[i])
      q = pi[q - 1];
    if (pattern[q] == text[i])
      q++;
    if (q == m) {
      matches.push_back(i - m + 1);
      q = pi[q - 1];
    }
  }
  return matches;
}

int main() {
  int needle_length;
  string needle, haystack;
  while (cin >> needle_length) {
    cin >> needle;
    cin >> haystack;
    vector<int> occurrences = kmp(haystack, needle);
    if (occurrences.empty()) {
      cout << endl;
    } else {
      for (int i = 0; i < occurrences.size(); i++)
        cout << occurrences[i] << endl;
    }
  }
  return 0;
}